Added timely vote credits proposal. (#33)
solana-foundation/solana-improvement-documents: proposals/0033-timely-vote-credits.md

over time.  It is estimated that about 7.5 million stake, or roughly 2% of all
stake on the network, is using an intentional vote lagging strategy.

## Alternatives Considered

It's hard to conceive an alternative to this proposal since the problem can
only be addressed via economic incentive to reduce vote lagging.  The only
alternatives would be in the specific characteristics of how credits are
awarded based on vote latency; several parameterizations were considered and
this was written about in an early description of this issue:

https://www.shinobi-systems.com/timely_voting_proposal

This website is quite out of date however; the parameterizations that were
proposed there ("4 16 1", see later in this document for a description of what
these numbers mean) has been long since superceded by different values ("2 8
1").  This is because the overall performance of the Solana mainnet cluster
has matured to a degree where "tighter timings" are tolerable.

## New Terminology

**Vote Latency** is defined as the difference, in slots, between the slot
being voted on within a vote transaction, and the slot in which the vote

> transaction itself lands.  The minimum possible vote latency for any voted-on

Reviewer: > and the slot in which the vote transaction itself lands

I have some concern about this w.r.t. the proposed grace period. Currently, the leader scheduler gives leaders 4 slots at a time to be leader. The leader chooses which vote transactions get packed into those blocks.

As a malicious operator, I now have some incentive to not pack competing validator's votes; by delaying them, it reduces their rewards, and makes my validator appear more attractive to potential stakers.

I think with the current system, there's no real incentive to do this because it doesn't matter if I don't pack them, the next leader will and they'll get the same reward. And now I've only lost out on the transaction fees myself. 

I wonder if the grace period should be at least 4 slots to avoid this? Even so, I can imagine delaying processing votes until my 4th slot for the same reason while still earning the fees.

Haven't though through it entirely, but a potential way to address this is by making vote credits based on slot latency mod 4 (subject to change if leader schedule does).

Author: In doing so validators will be taking a risk of loss of block rewards: deferring any transactions until a later block will miss out on block rewards if those later blocks are skipped.  In addition, deferring votes to later blocks increases the chances that those blocks will be skipped as votes which could have landed earlier to "strengthen" the fork will land later instead, giving greater chance for a competing fork to win.

Both of these risks are anti-incentive for the validator to intentionally defer votes.

The effect that a validator can have on its "competitors" by deferring votes is very minor: a validator only gets approximately its stake weight in slots per epoch.  I don't think this has ever risen much above 3% since the early days of mainnet-beta, and anything above 2% is pretty rare.

Practically speaking this means that a validator could at worst reduce about 2% of a competitor's votes by 2 slots, or assuming an average of 7.5 credits per slot, a reduction from 7.5 to 5.5 credits for 2% of the target validators slots.  This is a reduction in 0.5% of credits -- and at the cost of greater risk of lost block rewards to the offending validator.

This issue was discussed in other forums previously and the conclusion was that it's not an issue large enough to counterbalance the benefits of the proposal.

But I'd be happy to hear your further thoughts on this.

Reviewer 2: > In addition, deferring votes to later blocks increases the chances that those blocks will be skipped as votes which could have landed earlier to "strengthen" the fork will land later instead, giving greater chance for a competing fork to win.

Adding to this: 

Votes for blocks on different forks cannot land (they will fail as invalid tx in vote program), so a leader can only censor votes for their own fork - which as zan mentioned hurts their chance of being selected in the heaviest fork calculation. I imagine a competing validator would have a sizable stake so this disincentives such behavior. 

At best a leader could censor tiny validators which would have a minimal reward for loss of tx fees. Based on the numbers zan cited this seems unlikely to be beneficial. 

I think if any such adversarial block packing strategy emerges successfully the best way to counteract it would be to tweak the ratio of block rewards/vote tx fees/vote credits

When a Vote or VoteStateUpdate instruction is received by a validator, it uses
the Clock sysvar to identify the slot in which that instruction has landed.
For any newly voted on slot within that Vote or VoteStateUpdate transaction,
the validator will record the vote latency of that slot as (voted_in_slot -
voted_on_slot).

These vote latencies are stored per vote in a u8 as part of the on chain vote
state.

The credits awarded for a rooted slot are calculated using the latency value
for the corresponding vote, using a formula that awards latencies of 1 - 2
slots 8 credits, with a 1 credit reduction for each vote latency after 2.
Rooted slots will always be awarded a minimum credit of 1 (never 0) so that
very old votes, possibly necessary in times of network stress, are not
discouraged.

The implementation of the above is fairly simple for Vote transactions,
because Vote transactions include slot numbers voted on, and that list of
slots is already pruned of already-voted-on slots; so each newly voted-on slot
that survives this pruning can have its latency calculated simply.

## Impact

1. Any code which deserialized vote account state will require a recompile

>    with the newest Solana SDK in order to be able ot deserialize the new vote

Reviewer: typo:

```suggestion
   with the newest Solana SDK in order to be able to deserialize the new vote
```

Author: Will fix, thanks.

Reviewer 3: Ok

   SOL.  The true value will certainly be less than this because some vote
   accounts will already be above the new minimum.  In any case, the author of
   the SIMD will send any additional SOL needed to increase rent exempt
   minimum of all vote accounts, in the weeks leading up to the enablement of
   this feature.  By the time the feature is enabled, the vast majority of
   vote accounts will be funded to the new rent exempt minimum.

   For any vote account not sufficiently funded, the old VoteState version
   will be written.  Since the old VoteState does not record latencies, vote
   account deserialization will assign the default value of 0 to the Lockout
   latency field, which will result in the minimum vote credits awarded for
   each vote for that vote account.  This provides economic incentive to
   validators to sufficiently fund their vote accounts to allow the updated
   VoteStateVersion to be serialized into them.

   To summarize the issues surrounding the Vote account updates:

   - A vote account will only be updated to the new version after the feature
     is enabled, and only if the vote account includes enough SOL to satisfy

>      the new rent exempt minimum.  - If a vote account is not updated, then it

Reviewer: is this intended to be 2 bullet points?

Author: Will fix, got messed up in recent auto-fill

period slots, every additional slot of latency would result in a 10%
reduction in vote credits earned as each subsequent slot earns 1 credit less
out of a maximum possible 10 credits.

Again, historical voting data was analyzed over many epochs and the
conclusion drawn was that a maximum credits of 8 is the largest value that
can be used and still have a noticeable effect on known laggers.  Values
higher than that result in such a small penalty for each slot of lagging
that intentional lagging is still too profitable.  Lower values are even
more punishing to intentional lagging; but an attempt has been made to
conservatively choose the highest value that produces noticeable results.

When a Vote or VoteStateUpdate instruction is received by a validator, it
uses the Clock sysvar to identify the slot in which that instruction has
landed.  For any newly voted on slot within that Vote or VoteStateUpdate
transaction, the validator will record the vote latency of that slot as
(voted_in_slot - voted_on_slot).

These vote latencies are stored in u8 field of the Lockout data structure
named *latency*.

The credits awarded for a rooted slot are calculated using the latency value

> stored in the latency field the slot, using a formula that awards latencies

Reviewer 2: ```suggestion
for the corresponding vote, using a formula that awards latencies
```

of 1 - 2 slots 8 credits, with a 1 credit reduction for each vote latency
after 2.  Rooted slots will always be awarded a minimum credit of 1 (never
0) so that very old votes, possibly necessary in times of network stress,
are not discouraged.

> The implementation of the above is fairly simple for Vote transactions,

Reviewer 2: The remaining part of this section is not needed as `rooted slot` is well defined for both Vote and VoteStateUpdate instructions and was designed as part of the initial specs/proposal. Just makes it more complex re-explaining the rooting mechanism already present in the vote program.

Author: Also it allows me to refer to existing implementation to give assurance that the melding of the new implementation into the old won't cause problems, and to describe why that is true.  But again I will remove it.

VoteStateUpdate tx landed in) and this latency is stored in VoteState for
that slot.

The code to handle this is more complex, because VoteStateUpdate may include
removal of slots that expired as performed by the voting validator, in
addition to slots that have been rooted and new slots added.  However, the
assumptions that are needed to handle VoteStateUpdate with timely vote
credits are already guaranteed by existing VoteStateUpdate correctness
checking code:

The existing VoteStateUpdate processing code already ensures that (1) only
roots slots that could actually have been rooted in the transition from the
old VoteState to the new VoteState, so there is no danger of over-counting
credits (i.e. imagine that a 'cheating' validator "pretended" that slots
were rooted by dropping them off of the back of the new VoteState before
they have actually achieved 32 confirmations; the existing logic prevents
this).

The existing VoteStateUpdate processing code already ensures that (2) new
slots included in the new VoteState are only slots after slots that have
already been voted on in the old VoteState (i.e. can't inject new slots in
the middle of slots already voted on).

## Impact

> 1. Any code which deserialized vote account state will require a recompile

Reviewer 2: I think instead for this bullet point you should outline the feature flag for upgrading the new vote state format (and implications on any rent balance) as well as the non impact on the tower.

Reviewer 2: Actually I see you talked about it down below, maybe just a note here that the on chain vote state format will change and see the `Backwards Compatibility` section for the upgrade procedure.

   with the newest Solana SDK in order to be able ot deserialize the new
   vote account state.

2. There will be a strong anti-incentive for vote lagging.  Validators who
   currently delay their votes in order to reduce their chances of being
   locked out from voting, will no longer do so.

3. The Solana CLI will have been updated to show vote latencies for each
   vote; any context in which the Solana CLI output is observed by users
   or other tooling will need to adapt to the additonal data being
   displayed.

## Security Considerations

The are no security considerations; however there are changes to incentive
structure for voting which may result in differing vote behavior becoming
more optimal with regards to earning credits.

> To maximize vote credits, a validator must vote on all slots that end up

Reviewer 2: This paragraph and the 2 under it would be better in the motivation section, not really security considerations.

Author: I put that discussion under Security Considerations because are people who believe (and have said so in public discussion) that voting incentives can impact security; so here I lay out the case for why the new voting incentives do not have security implications.

being rooted by the cluster, and no slots that are not rooted.  This means
always "picking the right fork" whenever there are forks.  It is the drive
to always vote on the correct fork that causes validators to choose a vote
lagging strategy in the first place; they can earn more credits if they
reduce their chances of voting on "the wrong fork" by waiting until
sufficient votes from other validators have landed on a given slot before
voting on it.

Therefore the optimal vote strategy currently is to wait until a slot is
rooted (i.e. has received 31 confirmations) before voting on it.  In this
way, perfect vote credits would be earned.  However, this optimal voting
strategy, if employed cluster-wide, would halt the cluster as no validators
would vote, all of them waiting for others to vote first.

It's possible to get very close to this optimal strategy by waiting for some
number of confirmations less than 31, with the number of incidents of
picking the wrong fork increasing as the waiting time is decreased.  The
current "vote laggers" have done this, with varying degrees of skill.

> After the timely vote credits change, the dynamics will change; it will no

Reviewer 2: These three paragraphs would be better in the impact section, the impact on the new optimal voting strategy.

Author: Again the "Security Considerations" section was meant to specifically address issues that have already been brought up several times in validator meetings to discuss this feature; having identified that as a point that people generally have concern about, I thought it best to address that topic very clearly here.

   - If a vote account is not updated, then it will continue to work, but
     because its vote latencies will not be recorded, and will default to
     0, the vote account will earn minimum credits for votes until
     updated

   There is no danger of breaking vote processing; the only danger is to
   validators who do not supply sufficient rent exempt minimum, and that
   danger is limited to reduced earnings through lower vote credits.  There
   is no risk to cluster consensus.

2. A second change will be made, with a separate feature, that will enable
   the recording of the latency values for votes in the latency field of
   Lockout, and the subsequent use of this latency value to compute vote
   credits awarded.

   Once this feature is enabled, vote accounts will earn credits based on
   their vote latencies.  The maximum possible credits earnings per epoch
   will increase from 432,000 to 3,456,000.  Each validator will continue to
   earn vote rewards based on their fraction of total vote credits earned
   across the cluster, and validators which earn more credits through
   reduced vote latencies will earn a larger percentage of the total
   rewards.

> ## A Note About the Tower

Reviewer 2: I think this section can be folded into the previous, talking about the backwards compatibility of the tower.

Author: It can be, but why?  It's meant to provide documentation about a specific and complex and confusing aspect to the implementation.

Reviewer 2: SIMDs are client agnostic, any references to the Solana Labs implementation are better suited for the PR in that repo

Reviewer 2: SIMDs are client agnostic, any references to the Solana Labs implementation are better suited for the PR in that repo
